codalab/lib/upload_manager.py

```diff
@@ -258,3 +326,168 @@ def cleanup_existing_contents(self, bundle):
         bundle_update = {'data_hash': None, 'metadata': {'data_size': 0}}
         self._bundle_model.update_bundle(bundle, bundle_update)
         self._bundle_model.update_user_disk_used(bundle.owner_id)
+
+    def get_bundle_sas_token(self, path, **kwargs):
+        """
+        Get SAS token with write permission. Used for bypass server uploading.
+        """
+        return (
+            parse_linked_bundle_url(path)
+            .bundle_path_sas_url(permission='rw', **kwargs)
+            .split('?')[-1]
+        )
+
+    def get_index_sas_token(self, path, **kwargs):
+        """
+        Get SAS token of the index file with read and write permission. Used for uploading.
+        """
+        return (
+            parse_linked_bundle_url(path)
+            .index_path_sas_url(permission='rw', **kwargs)
+            .split('?')[-1]
+        )
+
+    def get_bundle_signed_url(self, path, **kwargs):
```

Comment:
What's the difference between a signed url and sas url?

Reply:
The signed url is for GCS, and the SAS URL is for Azure